apollo/smoothie.py

```diff
@@ -58,70 +24,182 @@ def connect(self):
                 loop=self.loop
             )
         except OSError as e:
-            logger.log("foo")
+            logger.error("Did not connect")
             raise e
 
         self.reader = reader
         self.writer = writer
         return (yield from self.smoothie_handshake())
 
+    def get_delimited_gcode(self, gcode):
+        """
+        # Add delimiter to gcode for Smoothie board
+        """
+        return gcode.strip() + '\r\n'
+
     @asyncio.coroutine
     def smoothie_handshake(self):
-        smoothie_ok = [False, False]
-
-        while not (smoothie_ok[0] and smoothie_ok[1]):
-            data = yield from self._read()
-            if data == 'Smoothie':
-                smoothie_ok[0] = True
-            elif data == 'ok':
-                smoothie_ok[1] = True
+        first_message = yield from self._read()
+        second_message = yield from self._read()
+        return (first_message == 'Smoothie') and (second_message == 'ok')
 
-        return smoothie_ok[0] and smoothie_ok[1]
+    @asyncio.coroutine
+    def write_and_drain(self, message):
+        try:
+            self.writer.write(message.encode('utf-8'))
+            yield from self.writer.drain()
+        except Exception as exc:
+            logger.error('Exception in write_and_drain: {}'.format(exc))
 
     @asyncio.coroutine
-    def send(self, gcode, response_handler):
-        self.writer.write(gcode.encode()) 
-        yield from self.writer.drain()
+    def send(self, gcode, response_handler=None):
+        # Skip if command is EMERGENCY STOP or RESET FROM HALT
+        if not (gcode == 'M112' or gcode == 'M999'):
+            yield from self.turn_on_feedback()
 
-        if response_handler:
-            response_handler.send('start', gcode)
+        delimited_gcode = self.get_delimited_gcode(gcode)
 
-        done = False
-        res = []
+        yield from self.write_and_drain(delimited_gcode)
 
-        while not done:
-            data = (yield from self._read())
-            if data == None:
-                done = True
-                break
+        if response_handler and asyncio.iscoroutine(response_handler):
+            yield from response_handler('start', gcode)
 
-            if response_handler:
-                response_handler.send(data)
+        is_gcode_done = False
+        json_result = None
 
-            if not self.feedback:
-                done = True
+        counter = 0
+        while not is_gcode_done:
+            counter += 1
+            if counter > 9:
+                break
+                raise SmoothieMaxReadsException(gcode) 
```

Comment:
I'm not sure how this raise exception is reached with the break in front of it...